### PR TITLE
A way forward for pointer equality in const eval

### PR DESCRIPTION
**Pull request by oli-obk**
r? @varkor on the first commit and @RalfJung on the second commit

cc #53020

### CODE REVIEW

File: src/test/ui/const-generics/raw-ptr-const-param-deref.stderr [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
    |
-LL | #![feature(const_generics, const_compare_raw_pointers)]
+LL | #![feature(const_generics)]
    |            ^^^^^^^^^^^^^^
    |
    = note: `#[warn(incomplete_features)]` on by default
    = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
 
-warning: 1 warning emitted
+error: using raw pointers as const generic parameters is unstable
```
Comment (varkor) [inline src/test/ui/const-generics/raw-ptr-const-param-deref.stderr:10]: Presumably we should be telling the user that this is forbidden, rather than simply unstable. Other than that, this commit looks fine to me.

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -1941,6 +1941,16 @@ extern "rust-intrinsic" {
     ///
     /// Perma-unstable: do not use.
     pub fn miri_start_panic(payload: *mut u8) -> !;
+
+    /// See documentation of `<*const T>::guaranteed_eq` for details.
+    #[rustc_const_stable(feature = "const_raw_ptr_comparison", since = "1.46")]
+    #[cfg(not(bootstrap))]
+    pub fn ptr_guaranteed_eq<T>(ptr: *const T, other: *const T) -> bool;
```
Comment (oli-obk) [inline src/libcore/intrinsics.rs:1948]: The reason I chose `guaranteed` instead of `maybe` (and invert the logic) is that I explained the `maybe` variant to someone and got recommended that I explain it this way, because it requires less negative reasoning.

File: src/libcore/ptr/const_ptr.rs [hunk 3]
```diff
@@ -295,6 +295,62 @@ impl<T: ?Sized> *const T {
         intrinsics::ptr_offset_from(self, origin)
     }
 
+    /// Returns whether two pointers are guaranteed equal. Returns `true` whenever the
+    /// pointers are known to be equal. If the pointers are equal, but the platform
+    /// isn't sure the pointers are equal, this function returns `false`.
+    ///
+    /// This function is the mirror of [`maybe_ne`], but not its inverse. There are pointer
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:302]: Spurious reference to `maybe_ne`? (There are a few of those.)

File: src/libcore/ptr/const_ptr.rs [hunk 4]
```diff
@@ -295,6 +295,62 @@ impl<T: ?Sized> *const T {
...
+    /// This function is the mirror of [`guaranteed_ne`], but not its inverse. There are pointer
+    /// comparisons for which both functions return false.
+    ///
+    /// [`guaranteed_ne`]: #method.guaranteed_ne
+    ///
+    /// # Safety
+    ///
+    /// This function may not be used to change the behaviour of any code. It may
+    /// only be used for short circuiting operations out of performance reasons.
+    /// It is unsound for this function to be used to allow safe code to observe
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:311]: For this to be unsound, it has to violate some soundness condition. Our usual soundness condition is "you may not cause UB from safe code", which is unaffected by this. So I am not sure if soundness is the right term here. Your new soundness condition (which IMO requires t-lang FCP on its own) also seems to involve relational reasoning, which is concerning -- I'd prefer for soundness to be a unary property, i.e. one that can be evaluated by looking at individual execution traces (as opposed to a property that relates multiple execution traces).

File: src/libcore/ptr/const_ptr.rs [hunk 5]
```diff
@@ -295,6 +295,62 @@ impl<T: ?Sized> *const T {
...
+    /// the return value of this function. You must be able to assume that the
+    /// function always returns `false` and your logic must still be sound.
+    /// ```
+    #[unstable(feature = "const_raw_ptr_comparison", issue = "53020")]
+    #[rustc_const_stable(feature = "const_raw_ptr_comparison", since = "1.46")]
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:316]: Wait, this is `rustc_const_stable`? That doesn't seem right.

File: src/librustc_mir/transform/check_consts/ops.rs [hunk 6]
```diff
@@ -268,18 +268,24 @@ impl NonConstOp for Panic {
...
             span,
-            &format!("comparing raw pointers inside {}", ccx.const_kind()),
-        )
-        .emit();
+            "pointers cannot be compared in a meaningful way during const eval.",
+        );
+        err.note(
+            "It is conceptually impossible for const eval to know in all cases whether two \
+             pointers are equal. While sometimes it is clear (the address of a static item \
+             is never equal to the address of another static item), comparing an integer \
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/ops.rs:279]: Except if one of the two static items is zero-sized...

File: src/librustc_mir/transform/check_consts/ops.rs [hunk 7]
```diff
@@ -268,18 +268,24 @@ impl NonConstOp for Panic {
...
+             address with any allocation's address is impossible to do at compile-time.",
+        );
+        err.note(
+            "That said, there's the `ptr_guaranteed_eq` intrinsic which returns `false` for all \
+             comparisons where CTFE isn't sure whether two addresses are equal. The mirror \
+             intrinsic `ptr_guaranteed_eq` returns `false` for all comparisons where CTFE isn't \
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/ops.rs:285]:
```suggestion
             intrinsic `ptr_guaranteed_ne` returns `false` for all comparisons where CTFE isn't \
```
But also your explanation is using double-negation ("false if not sure"), wouldn't it be easier to describe this as "true if it will definitely be equal, and false if we cannot be sure"?

File: src/test/ui/consts/const-eval/const_raw_ptr_ops.stderr [hunk 8]
```diff
@@ -1,44 +1,20 @@
...
    |
 LL | const X: bool = unsafe { &1 as *const i32 == &2 as *const i32 };
-   | -------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^---
-   |                          |
-   |                          "pointer arithmetic or comparison" needs an rfc before being allowed inside constants
+   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
-   = note: `#[deny(const_err)]` on by default
+   = note: It is conceptually impossible for const eval to know in all cases whether two pointers are equal. While sometimes it is clear (the address of a static item is never equal to the address of another static item), comparing an integer address with any allocation's address is impossible to do at compile-time.
+   = note: That said, there's the `ptr_guaranteed_eq` intrinsic which returns `false` for all comparisons where CTFE isn't sure whether two addresses are equal. The mirror intrinsic `ptr_guaranteed_eq` returns `false` for all comparisons where CTFE isn't sure whether two addresses are inequal.
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/const_raw_ptr_ops.stderr:8]: Wow, so much text. This will totally explode people's command-line window.^^

File: src/librustc_mir/interpret/intrinsics.rs [hunk 9]
```diff
@@ -291,6 +291,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let offset_ptr = ptr.ptr_wrapping_signed_offset(offset_bytes, self);
                 self.write_scalar(offset_ptr, dest)?;
             }
+            sym::ptr_guaranteed_eq | sym::ptr_guaranteed_ne => {
+                // FIXME: return `true` for at least some comparisons where we know equality and
+                // inequality have meaning.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:296]:
```suggestion
                // FIXME: return `true` for at least some comparisons where we can determine (in)equality
                // at compile time.
```
"have meaning" doesn't really have much meaning here IMO. ;)

File: src/libcore/ptr/const_ptr.rs [hunk 10]
```diff
@@ -295,6 +295,62 @@ impl<T: ?Sized> *const T {
         intrinsics::ptr_offset_from(self, origin)
     }
 
+    /// Returns whether two pointers are guaranteed equal. Returns `true` whenever the
+    /// pointers are known to be equal. If the pointers are equal, but the platform
+    /// isn't sure the pointers are equal, this function returns `false`.
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:300]: What is "the platform"?

I'd say something like

> Returns whether two pointers are definitely equal. In some evaluation contexts (e.g., compile-time evaluation),
> it is not always possible to determine equality of two pointers, so this function may spuriously return
> `false` for pointers that later actually turn out to be equal.

File: src/libcore/ptr/const_ptr.rs (hunk 4 above)
Comment (oli-obk) [reply src/libcore/ptr/const_ptr.rs]: Right, I adjusted the comment and made the function safe

File: src/libcore/ptr/const_ptr.rs (hunk 5 above)
Comment (oli-obk) [reply src/libcore/ptr/const_ptr.rs]: made it unstable again so we don't accidentally stabilize

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs]: added that caveat

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 7 above)
Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs]: I improved the explanation (and got rid of the double negative) lmk what you think

File: src/librustc_mir/interpret/intrinsics.rs (hunk 9 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/intrinsics.rs]: I rewrote the FIXME

File: src/libcore/ptr/const_ptr.rs (hunk 10 above)
Comment (oli-obk) [reply src/libcore/ptr/const_ptr.rs]: I took your text and adjusted it a bit.

File: src/libcore/intrinsics.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/intrinsics.rs:1948]: Seems reasonable, this also means our only example usecase does not need a negation. :)

File: src/libcore/ptr/const_ptr.rs [hunk 11]
```diff
@@ -295,6 +295,64 @@ impl<T: ?Sized> *const T {
         intrinsics::ptr_offset_from(self, origin)
     }
 
+    /// Returns whether two pointers are guaranteed equal. Returns `true` whenever the
+    /// pointers are known to be equal.
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:299]: This intro still sounds rather odd, it is basically saying the same thing twice, or at least it sounds like it does.

File: src/libcore/ptr/const_ptr.rs [hunk 12]
```diff
@@ -295,6 +295,64 @@ impl<T: ?Sized> *const T {
...
+    /// In some contexts (e.g., compile-time evaluation),
+    /// it is not always possible to determine equality of two pointers, so this function may
+    /// spuriously return false for pointers that later actually turn out to be equal.
+    /// But when it returns true, the pointers are guaranteed to be equal.
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:303]:
```suggestion
    /// spuriously return `false` for pointers that later actually turn out to be equal.
    /// But when it returns `true`, the pointers are guaranteed to be equal.
```

Same for `guaranteed_ne`.

File: src/libcore/ptr/const_ptr.rs [hunk 13]
```diff
@@ -295,6 +295,64 @@ impl<T: ?Sized> *const T {
...
+    /// This function is the mirror of [`guaranteed_ne`], but not its inverse. There are pointer
+    /// comparisons for which both functions return false.
+    ///
+    /// [`guaranteed_ne`]: #method.guaranteed_ne
+    ///
+    /// The return value may change depending on the compiler version and unsafe code may not
+    /// rely on the result of this function for soundness. It is suggested to only use this function
+    /// for performance optimizations where spurious `false` return values by this function do not
+    /// affect the outcome, but just the performance.
+    /// During runtime this function compiles to a direct pointer comparison.
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:314]:
```suggestion
    /// At runtime this function behaves like `self == other`.
```
But this sounds like it belongs in the very first paragraph, not down here.

Also for full transparency I would add something like this:

> The consequences of using this method to make runtime and compile-time code behave differently have not been explored. This method should *not* be used to introduce such differences, and it should also not be stabilized before we have a better understand of this issue.

File: src/librustc_mir/interpret/intrinsics.rs [hunk 14]
```diff
@@ -291,6 +291,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let offset_ptr = ptr.ptr_wrapping_signed_offset(offset_bytes, self);
                 self.write_scalar(offset_ptr, dest)?;
             }
+            sym::ptr_guaranteed_eq | sym::ptr_guaranteed_ne => {
+                // FIXME: return `true` for at least some comparisons where we know equality and
+                // inequality will always be equal to the runtime equivalent.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:296]: "equality equal to the equivalent"? Lots of equals. ;)
```suggestion
                // FIXME: return `true` for at least some comparisons where we can reliably
                // determine the result of runtime (in)equality tests at compile-time.
```

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:296]: I think I suggested something like this before, and I still think "reliably determine runtime result" is the clearest way to express this, which is why I suggested it both here and for the doc comment. Not sure why you did not like it here?

File: src/librustc_mir/transform/check_consts/ops.rs [hunk 15]
```diff
@@ -268,18 +268,24 @@ impl NonConstOp for Panic {
...
+        let mut err = ccx.tcx.sess.struct_span_err(
             span,
-            &format!("comparing raw pointers inside {}", ccx.const_kind()),
-        )
-        .emit();
+            "pointers cannot be compared in a meaningful way during const eval.",
+        );
+        err.note(
+            "It is conceptually impossible for const eval to know in all cases whether two \
+             pointers are equal. While sometimes it is clear (the address of a non-zst static item \
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/ops.rs:278]:
```suggestion
             pointers are equal at runtime. While sometimes it is clear (the address of a non-zst static item \
```
Also shouldn't this be broken at 80 or 100 characters or so?

File: src/librustc_mir/transform/check_consts/ops.rs [hunk 16]
```diff
@@ -268,18 +268,24 @@ impl NonConstOp for Panic {
...
+             is never equal to the address of another non-zst static item), comparing an integer \
+             address with any allocation's address is impossible to do at compile-time.",
+        );
+        err.note(
+            "That said, there's the `ptr_guaranteed_eq` intrinsic which returns `true` for all \
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/ops.rs:283]: This part of the message should not appear on stable compilers.

File: src/librustc_mir/transform/check_consts/ops.rs [hunk 17]
```diff
@@ -268,18 +268,24 @@ impl NonConstOp for Panic {
...
+             comparisons where CTFE is sure that two addresses are equal. The mirror \
+             intrinsic `ptr_guaranteed_eq` returns `true` for all comparisons where CTFE is \
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/ops.rs:285]:
```suggestion
             intrinsic `ptr_guaranteed_ne` returns `true` for all comparisons where CTFE is \
```
I think I pointed this out before?^^

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 15 above)
Comment (RalfJung) [reply src/librustc_mir/transform/check_consts/ops.rs:278]: Oh and do we usually capitalize ZST or not?

Comment (RalfJung) [reply src/librustc_mir/transform/check_consts/ops.rs:278]: Honestly, this is probably too much text. I think it makes more sense to just say "we cannot reliably determine whether two pointers are equal at compiletime, see URL for further information".

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs:282]: With rustfmt that isn't enforced anymore, I fixed that

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 17 above)
Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs]: oops, I thought I fixed this

File: src/libcore/ptr/const_ptr.rs [hunk 18]
```diff
@@ -295,23 +295,26 @@ impl<T: ?Sized> *const T {
         intrinsics::ptr_offset_from(self, origin)
     }
 
-    /// Returns whether two pointers are guaranteed equal. Returns `true` whenever the
-    /// pointers are known to be equal.
+    /// Returns whether two pointers are guaranteed equal.
+    /// At runtime this function behaves like `self == other`.
     /// In some contexts (e.g., compile-time evaluation),
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs]:
```suggestion
    /// Returns whether two pointers are guaranteed equal.
    /// 
    /// At runtime this function behaves like `self == other`.
    /// However, in some contexts (e.g., compile-time evaluation),
```
Yeah now I am nitpicking. ;)

File: src/libcore/ptr/const_ptr.rs [hunk 19]
```diff
@@ -295,23 +295,26 @@ impl<T: ?Sized> *const T {
...
     /// [`guaranteed_ne`]: #method.guaranteed_ne
     ///
     /// The return value may change depending on the compiler version and unsafe code may not
     /// rely on the result of this function for soundness. It is suggested to only use this function
     /// for performance optimizations where spurious `false` return values by this function do not
     /// affect the outcome, but just the performance.
-    /// During runtime this function compiles to a direct pointer comparison.
+    /// The consequences of using this method to make runtime and compile-time code behave
+    /// differently have not been explored. This method should not be used to introduce such
+    /// differences, and it should also not be stabilized before we have a better understand
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:316]:
```suggestion
    /// differences, and it should also not be stabilized before we have a better understanding
```

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 15 above)
Comment (RalfJung) [reply src/librustc_mir/transform/check_consts/ops.rs:282]: I meant the output, not the source.

Comment (RalfJung) [reply src/librustc_mir/transform/check_consts/ops.rs:282]: Also even with your latest changes this is still a wall of text -- it is *by far* the largest diagnostic message I have ever seen. Wouldn't it make sense to just move all that information behind some URL? In particular the non-zst static example seems like way too much detail.

File: src/librustc_mir/interpret/intrinsics.rs (hunk 14 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/intrinsics.rs]: Sorry, your review was good, I just didn't read it properly :/

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 15 above)
Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs:298]: I can move it behind `--teach`

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs:298]: Sorry I only saw half of the comments here. Github didn't show me anything but the initial comment before my last comment when I wrote it

Comment (RalfJung) [reply src/librustc_mir/transform/check_consts/ops.rs:298]: > I can move it behind --teach

I never used that flag, so -- maybe? But the more detailed explanations are AFAIK usually put in markdown file or so?

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs]: for now all the info is behind the linked issue, so let's just go with that

File: src/librustc_mir/transform/check_consts/ops.rs [hunk 20]
```diff
@@ -287,26 +287,12 @@ impl NonConstOp for RawPtrComparison {
     fn emit_error(&self, ccx: &ConstCx<'_, '_>, span: Span) {
         let mut err = ccx.tcx.sess.struct_span_err(
             span,
-            "pointers cannot be compared in a meaningful way during const eval.",
+            "pointers cannot be reliably compared way during const eval.",
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/ops.rs:290]:
```suggestion
            "pointers cannot be reliably compared during const eval.",
```

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 15 above)
Comment (RalfJung) [reply src/librustc_mir/transform/check_consts/ops.rs]: LGTM. Might be worth updating the OP of that issue with some of the text we worked on here.

File: src/libcore/intrinsics.rs [hunk 21]
```diff
@@ -11,7 +11,7 @@
 //! In order to make an intrinsic usable at compile-time, one needs to copy the implementation
 //! from https://github.com/rust-lang/miri/blob/master/src/shims/intrinsics.rs to
 //! `librustc_mir/interpret/intrinsics.rs` and add a
-//! `#[rustc_const_unstable(feature = "foo", issue = "01234")]` to the intrinsic.
+//! `#[rustc_const_unstable(feature = "foo2", issue = "01234")]` to the intrinsic.
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs:14]: Hu?

Comment (oli-obk) [reply src/libcore/intrinsics.rs:14]: This is part of 84f1d73

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 20 above)
Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs]: oops, fixed

File: src/libcore/intrinsics.rs (hunk 21 above)
Comment (RalfJung) [reply src/libcore/intrinsics.rs:14]: But why does that require renaming `foo` to `foo2`? (Maybe `bar` would be more on-brand.^^)

Comment (oli-obk) [reply src/libcore/intrinsics.rs]: I fixed up the other use of `foo` so they are aligned.

File: src/libcore/ptr/const_ptr.rs [hunk 22]
```diff
@@ -295,6 +295,72 @@ impl<T: ?Sized> *const T {
...
+    #[inline]
+    #[cfg(not(bootstrap))]
+    pub const fn guaranteed_eq(self, other: *const T) -> bool
+    where
+        T: Sized,
+    {
+        intrinsics::ptr_guaranteed_eq(self, other)
+    }
+
+    /// Returns whether two pointers are guaranteed not equal.
```
Comment (RalfJung) [inline src/libcore/ptr/const_ptr.rs:331]: Is it just me or would "guaranteed to not be equal" or "guaranteed not to be equal" be grammatically more correct?

File: src/libcore/slice/mod.rs [hunk 23]
```diff
@@ -5983,6 +5985,50 @@ where
...
+    A: PartialEq<A> + Eq,
+{
+    default fn equal(&self, other: &[A]) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+
+        // While performance would suffer if `guaranteed_eq` just returned `false`
+        // for all arguments, correctness and return value of this function are not affected.
+        if self.as_ptr().guaranteed_eq(other.as_ptr()) {
```
Comment (RalfJung) [inline src/libcore/slice/mod.rs:6001]: I am curious, why did you duplicate the entire `impl` block instead of just a block inside this function?

**Review by RalfJung — APPROVED** [review]
LGTM from a Miri/const-eval perspective. I didn't check codegen or static const checks.

File: src/libcore/slice/mod.rs (hunk 23 above)
Comment (oli-obk) [reply src/libcore/slice/mod.rs:5967]: I just forgot that was possible. Fixed

File: src/libcore/ptr/const_ptr.rs (hunk 22 above)
Comment (oli-obk) [reply src/libcore/ptr/const_ptr.rs]: Yea, I also adjusted the text for `guaranteed_eq`

File: src/librustc_mir/transform/check_consts/ops.rs (hunk 15 above)
Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/ops.rs]: done

File: src/librustc_codegen_llvm/intrinsic.rs [hunk 24]
```diff
@@ -731,6 +731,18 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 return;
             }
 
+            "ptr_guaranteed_eq" | "ptr_guaranteed_ne" => {
+                let a = args[0].immediate();
+                let b = args[1].immediate();
+                let a = self.ptrtoint(a, self.type_isize());
+                let b = self.ptrtoint(b, self.type_isize());
```
Comment (nagisa) [inline src/librustc_codegen_llvm/intrinsic.rs:738]: Is there a non-obvious reason you `ptrtoint` before comparing? `icmp` can work fine with pointer types, as long as the pointee types match, which in this case they should.

File: src/libcore/ptr/const_ptr.rs [hunk 25]
```diff
@@ -295,6 +295,72 @@ impl<T: ?Sized> *const T {
...
+    /// spuriously return `false` for pointers that later actually turn out to be equal.
+    /// But when it returns `true`, the pointers are guaranteed to be equal.
+    ///
+    /// This function is the mirror of [`guaranteed_ne`], but not its inverse. There are pointer
+    /// comparisons for which both functions return `false`.
+    ///
+    /// [`guaranteed_ne`]: #method.guaranteed_ne
+    ///
+    /// The return value may change depending on the compiler version and unsafe code may not
+    /// rely on the result of this function for soundness. It is suggested to only use this function
```
Comment (Aaron1011) [inline src/libcore/ptr/const_ptr.rs:312]: Can unsafe code rely on a return value of `true` (similar to relying on `std::mem::needs_drop` returning `false`)?

**Review by nagisa — COMMENTED** [review]
r=me once there a comment explaining the ptrtoint or after the unnecessary instruction is removed.

File: src/libcore/ptr/const_ptr.rs (hunk 25 above)
Comment (RalfJung) [reply src/libcore/ptr/const_ptr.rs:312]: Yes, it can. The comment says 

> But when it returns `true`, the pointers are guaranteed to be equal.

I thought that was clear enough?

File: src/librustc_codegen_llvm/intrinsic.rs (hunk 24 above)
Comment (oli-obk) [reply src/librustc_codegen_llvm/intrinsic.rs:738]: oh. I didn't know that. I basically grabbed this off https://github.com/rust-lang/rust/blob/38bd83df88288f2f8d1fc2dd317189cac3825920/src/librustc_codegen_llvm/intrinsic.rs#L734-L748 which I'm not sure whether it needs that either.

Comment (RalfJung) [reply src/librustc_codegen_llvm/intrinsic.rs]: > which I'm not sure whether it needs that either.

It does. LLVM supports comparison of pointer types but not subtraction.

Comment (oli-obk) [reply src/librustc_codegen_llvm/intrinsic.rs]: So confusing. There's probably a good reason for that, but still... So many inconsistencies.

Comment (RalfJung) [reply src/librustc_codegen_llvm/intrinsic.rs]: FWIW, there's a plan to add a `psub` operation taht directly subtracts pointers. ;)

(There is no pointer addition, either. Instead there is `getelementptr` as the specific instruction to do pointer arithmetic. That makes a lot of sense IMO.)

File: src/libcore/slice/mod.rs [hunk 26]
```diff
@@ -5956,10 +5956,18 @@ where
...
         }
 
+        #[cfg(bootstrap)]
         if self.as_ptr() == other.as_ptr() {
             return true;
         }
 
+        // While performance would suffer if `guaranteed_eq` just returned `false`
+        // for all arguments, correctness and return value of this function are not affected.
+        #[cfg(not(bootstrap))]
```
Comment (bjorn3) [inline src/libcore/slice/mod.rs:5966]: Why was this changed? This function is not const.

File: src/librustc_mir/interpret/intrinsics.rs [hunk 27]
```diff
@@ -291,6 +291,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let offset_ptr = ptr.ptr_wrapping_signed_offset(offset_bytes, self);
                 self.write_scalar(offset_ptr, dest)?;
             }
+            sym::ptr_guaranteed_eq | sym::ptr_guaranteed_ne => {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/intrinsics.rs:294]: I just realized that this implementation will also be used by Miri, which is a bug. Can we somehow make it be used only during CTCE?

Comment (oli-obk) [reply src/librustc_mir/interpret/intrinsics.rs:294]: doesn't miri fall back to this only if it has no own impl? So we can just give miri an impl that behaves differently.

Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:294]: Miri prefers the CTFE impls, to make sure we test those properly.

I am working on a PR that adds these to Miri in a way that it prefers its own impls.

File: src/libcore/slice/mod.rs (hunk 26 above)
Comment (oli-obk) [reply src/libcore/slice/mod.rs:5966]: Not yet, we're working towards that. I have a local branch where I'm trying to run this in `-Zunleash-the-miri-inside-of-you` mode.

File: src/librustc_mir/interpret/intrinsics.rs (hunk 27 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics.rs:294]: Here it is: https://github.com/rust-lang/miri/pull/1459
